File: backend/endpoints/overview.py
from __future__ import annotations

from collections import defaultdict

from fastapi import APIRouter

from backend.data_loader import DATA
from backend.filters import apply_filters
from backend.helpers import dominant_sentiment, pct, sentiment_breakdown, week_key
from backend.models import FilterState

router = APIRouter()


@router.post("/overview")
def overview(req: FilterState):
    df, search_meta = apply_filters(req)
    crisis = DATA.get("crisis") or {}
    total = len(df)

    if total == 0:
        return {
            "kpis": {
                "total_posts": 0,
                "positive": 0,
                "negative": 0,
                "neutral": 0,
                "positive_pct": 0,
                "negative_pct": 0,
                "neutral_pct": 0,
                "unique_authors": 0,
                "total_entities": 0,
            },
            "crisis": {
                "level": crisis.get("level", "green"),
                "reason": crisis.get("reason", "No posts match current filters."),
                "negative_ratio": 0,
            },
            "sentiment_donut": [],
            "sentiment_trend": [],
            "search_meta": search_meta,
        }

    breakdown = sentiment_breakdown(df)
    pos, neg, neu = breakdown["positive"], breakdown["negative"], breakdown["neutral"]
    authors = df["author"].nunique() if "author" in df.columns else 0

    filtered_ids = set(df["post_id"].tolist())
    entity_count = 0
    seen_ent = set()
    for row in DATA.get("entities", []):
        if row.get("post_id") not in filtered_ids:
            continue
        for ent in row.get("entities") or []:
            k = ent["text"].lower()
            if k not in seen_ent:
                seen_ent.add(k)
                entity_count += 1

    weekly: dict[str, dict[str, int]] = defaultdict(lambda: {"positive": 0, "negative": 0, "neutral": 0})
    for _, row in df.iterrows():
        wk = week_key(row["created_utc"])
        lab = row.get("sentiment", "neutral")
        if lab in weekly[wk]:
            weekly[wk][lab] += 1

    trend = [{"week": w, **weekly[w]} for w in sorted(weekly.keys())]
    neg_ratio = neg / total if total else 0

    return {
        "kpis": {
            "total_posts": total,
            "positive": pos,
            "negative": neg,
            "neutral": neu,
            "positive_pct": pct(pos, total),
            "negative_pct": pct(neg, total),
            "neutral_pct": pct(neu, total),
            "unique_authors": int(authors),
            "total_entities": entity_count,
        },
        "crisis": {
            "level": crisis.get("level", "green"),
            "reason": crisis.get("reason", ""),
            "negative_ratio": round(neg_ratio, 3),
        },
        "sentiment_donut": [
            {"label": "positive", "count": pos, "color": "#22C55E"},
            {"label": "negative", "count": neg, "color": "#EF4444"},
            {"label": "neutral", "count": neu, "color": "#94A3B8"},
        ],
        "sentiment_trend": trend,
        "search_meta": search_meta,
    }

File: frontend/panels/overview.js
let donutChart = null;
let trendChart = null;

function renderOverview(data) {
  const k = data.kpis || {};
  const row = document.getElementById("kpi-row");
  row.innerHTML = `
    <div class="kpi-card"><div class="kpi-value">${k.total_posts ?? 0}</div><div class="kpi-label">Posts</div></div>
    <div class="kpi-card"><div class="kpi-value">${k.positive_pct ?? 0}%</div><div class="kpi-label">Positive</div></div>
    <div class="kpi-card"><div class="kpi-value">${k.negative_pct ?? 0}%</div><div class="kpi-label">Negative</div></div>
    <div class="kpi-card"><div class="kpi-value">${k.neutral_pct ?? 0}%</div><div class="kpi-label">Neutral</div></div>
    <div class="kpi-card"><div class="kpi-value">${k.unique_authors ?? 0}</div><div class="kpi-label">Authors</div></div>
    <div class="kpi-card"><div class="kpi-value">${k.total_entities ?? 0}</div><div class="kpi-label">Entities</div></div>
  `;

  const crisis = data.crisis || {};
  const badge = document.getElementById("crisis-badge");
  const level = (crisis.level || "green").toLowerCase();
  badge.className = `crisis-badge ${level}`;
  badge.innerHTML = `<i class="ti ti-alert-triangle"></i> ${level.toUpperCase()}: ${crisis.reason || ""}`;

  const crisisPanel = document.getElementById("crisis-content");
  crisisPanel.innerHTML = `
    <p><strong>Level:</strong> ${level}</p>
    <p>${crisis.reason || "No crisis data."}</p>
    <p><strong>Negative ratio (filtered):</strong> ${((crisis.negative_ratio || 0) * 100).toFixed(1)}%</p>
  `;

  renderDonut(data.sentiment_donut || []);
  renderTrend(data.sentiment_trend || []);
}

function renderDonut(slices) {
  const canvas = document.getElementById("chart-sentiment-donut");
  const panel = document.getElementById("panel-sentiment");
  if (!slices.length) {
    if (donutChart) { donutChart.destroy(); donutChart = null; }
    return;
  }
  if (donutChart) donutChart.destroy();
  donutChart = new Chart(canvas, {
    type: "doughnut",
    data: {
      labels: slices.map((s) => s.label),
      datasets: [{
        data: slices.map((s) => s.count),
        backgroundColor: slices.map((s) => s.color),
      }],
    },
    options: {
      onClick: (_, els) => {
        if (els.length && typeof openDrawer === "function") {
          const label = slices[els[0].index].label;
          openDrawer("sentiment", label);
        }
      },
      plugins: { legend: { position: "bottom" } },
    },
  });
}

function renderTrend(trend) {
  const canvas = document.getElementById("chart-sentiment-trend");
  if (!trend.length) {
    if (trendChart) { trendChart.destroy(); trendChart = null; }
    return;
  }
  if (trendChart) trendChart.destroy();
  trendChart = new Chart(canvas, {
    type: "bar",
    data: {
      labels: trend.map((t) => t.week),
      datasets: [
        { label: "Positive", data: trend.map((t) => t.positive), backgroundColor: "#22C55E" },
        { label: "Negative", data: trend.map((t) => t.negative), backgroundColor: "#EF4444" },
        { label: "Neutral", data: trend.map((t) => t.neutral), backgroundColor: "#94A3B8" },
      ],
    },
    options: {
      scales: { x: { stacked: true }, y: { stacked: true } },
      onClick: (_, els) => {
        if (els.length && typeof openDrawer === "function") {
          const week = trend[els[0].index].week;
          openDrawer("week", week);
        }
      },
    },
  });
}

function updateSearchBadge(meta) {
  const el = document.getElementById("search-badge");
  if (!meta || !meta.matched_entity) {
    el.hidden = true;
    return;
  }
  el.hidden = false;
  el.textContent = `Also matches entity: ${meta.matched_entity} (${meta.matched_entity_count || 0} posts)`;
  el.onclick = () => {
    document.getElementById("search-input").value = meta.matched_entity;
    if (typeof filterState !== "undefined") {
      filterState.search = meta.matched_entity;
      syncStateToURL();
      refreshDashboard();
    }
  };
}

function showPanelSkeleton(id) {
  const el = document.getElementById(id);
  if (el && !el.querySelector(".skeleton")) {
    const sk = document.createElement("div");
    sk.className = "skeleton";
    el.appendChild(sk);
  }
}

function clearPanelSkeleton(id) {
  document.querySelectorAll(`#${id} .skeleton`).forEach((n) => n.remove());
}
